### ISSUE TITLE
remove_training_nodes (Python Lib) Removes Identity Node Even When is Output

### ISSUE THREAD

**Issue by cancan101** [tensorflow/tensorflow#8181]
[`remove_training_nodes`](https://github.com/tensorflow/tensorflow/blob/64edd34ce69b4a8033af5d217cb8894105297d8a/tensorflow/python/framework/graph_util_impl.py#L246) removes `Identity` nodes even if that node is the specified graph output. 

For example if the `Identity` node was used to name and make more easily locatable the output of the graph.

**Comment by prb12** [comment]
@drpngx I believe you wrote this code?  Could you please comment.

**Comment by drpngx** [comment]
@petewarden who contributed the original code.

@cancan101 Could you send a simple, compact repro case?

Not sure, we have another case internally where replacing the [if clause](https://github.com/tensorflow/tensorflow/blob/64edd34ce69b4a8033af5d217cb8894105297d8a/tensorflow/python/framework/graph_util_impl.py#L310) fixed a problem.
```
      while input_name in names_to_splice:
        full_input_name = names_to_splice[input_name]
        input_name = re.sub(r"^\^", "", full_input_name)
      new_node.input.append(full_input_name)
```

**Comment by gonnet** [comment]
Fix submitted, but not quite sure when/how this will propagate here.

**Comment by drpngx** [comment]
Thanks! We try to push every day, hopefully it'll be there tomorrow. If you included `Fixes#8181` in your CL description, then it'll close this issue automatically when the fix is merged.

**Comment by gonnet** [comment]
Looks like the fix is in, but I don't seem to be able to close this myself...

**Comment by cancan101** [comment]
Can you link to PR?

**Comment by gonnet** [comment]
I think that would be https://github.com/tensorflow/tensorflow/pull/8402.

**Comment by drpngx** [comment]
In particular: https://github.com/tensorflow/tensorflow/pull/8402/commits/221cfa1e31be026a6f459410f0ed930befd8371d

Did that work for you @cancan101 ?

**Comment by cancan101** [comment]
Not sure, the issue I was having was that the last node (the output node) was of type Identity and that was getting pruned itself.

**Comment by drpngx** [comment]
Can you try with the updated code?

**Comment by gonnet** [comment]
Oh, then this may not fix your problem...

![no see that solution is for a different problem than the one i have](http://4.bp.blogspot.com/-KzOtz-8coJU/Uga2fN7SNmI/AAAAAAAAKI4/6QXtU2oXcJ4/s1600/ADTWO21.png)

**Comment by cancan101** [comment]
Here is a short example demonstrating the issue:
```python
import tensorflow as tf
from tensorflow.python.framework.graph_util_impl import remove_training_nodes

with tf.Graph().as_default() as g:
    input = tf.placeholder(tf.float32)
    output = tf.identity(input, name='output')
    graph_def = g.as_graph_def()
    print("Before:")
    print([n.name for n in graph_def.node])
    print("After:")    
    print([n.name for n in remove_training_nodes(graph_def).node])
```

which produces:
```
Before:
[u'Placeholder', u'output']
After:
[u'Placeholder']
```
In this case the signature of `remove_training_nodes` might need to be changed to list the set of nodes that should not be removed. 

Thoughts?

**Comment by girving** [comment]
@drpngx What's the status?

**Comment by drpngx** [comment]
We changed the `remove_training_nodes` to include a new parameter `protected_nodes`, which are kept unconditionally. @cancan101 would that work for you?

**Comment by cancan101** [comment]
That should work.

**Comment by drpngx** [comment]
Great!

**Comment by xiusir** [comment]
Maybe I have the same problem. TensorFlow VERSION  1.1.0
Has it been fixed in 1.1.0?

I specifed ["input", "output"] as the input and output nodes of my model  by tf.identity.
After frozen,  I can found "input", "output" in graph_def.node.
After optimized, "input" is still there, but "output" is lost.

# train.py
```python
def inference(images):
  images = tf.identity(images, 'input')
... ...
  softmax_linear = tf.identity(softmax_linear, 'output')
  return softmax_linear
```

# package.py -- Generate model file
```python
# -*- coding: utf-8 -*-
# Preparing a TF model for usage in Android
# pylint: disable=missing-docstring
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
import sys
import tensorflow as tf
from tensorflow.python.tools import freeze_graph
from tensorflow.python.tools import optimize_for_inference_lib

MODEL_NAME = 'carc34'
path='.'
input_graph_path = '%s/graph.pbtxt' % path
checkpoint_path = '%s/model.ckpt-100000' % path
input_saver_def_path = ""
input_binary = False
input_node_names = "input"
output_node_names = "output,softmax_linear/softmax_linear"
restore_op_name = "save/restore_all"
filename_tensor_name = "save/Const:0"
output_frozen_graph_name = 'frozen_'+MODEL_NAME+'.pb'
output_optimized_graph_name = 'optimized_'+MODEL_NAME+'.pb'
clear_devices = True

# Freeze the graph
freeze_graph.freeze_graph(input_graph_path, input_saver_def_path,
                          input_binary, checkpoint_path, output_node_names,
                          restore_op_name, filename_tensor_name,
                          output_frozen_graph_name, clear_devices, "")

# Optimize for inference
input_graph_def = tf.GraphDef()
with tf.gfile.Open(output_frozen_graph_name, "rb") as f:
    data = f.read()
    input_graph_def.ParseFromString(data)
    output_graph_def = optimize_for_inference_lib.optimize_for_inference(
                input_graph_def,
                input_node_names.split(","),  # an array of the input node(s)
                output_node_names.split(","), # an array of the output nodes
                tf.float32.as_datatype_enum)

# Save the optimized graph
f = tf.gfile.FastGFile(output_optimized_graph_name, "w")
f.write(output_graph_def.SerializeToString())
##tf.train.write_graph(output_graph_def, './', output_optimized_graph_name)   #

```

# show_nodes.py -- Print all nodes in frozen and optimized
```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
import os.path
import numpy as np
import tensorflow as tf
from tensorflow.python.framework import graph_util
from tensorflow.python.framework import tensor_shape
from tensorflow.python.platform import gfile
from tensorflow.python.util import compat

MODEL_OUTPUT_TENSOR_NAME = 'output'
with tf.Graph().as_default() as graph:
    model_filename = os.path.join('.', 'frozen_carc34.pb')
    with gfile.FastGFile(model_filename, 'rb') as f:
      graph_def = tf.GraphDef()
      graph_def.ParseFromString(f.read())
      ##for n in graph_def.node:
      ##  print (n.name)
      output_tensor = tf.import_graph_def(graph_def, name='', return_elements=[MODEL_OUTPUT_TENSOR_NAME])
      print ("Found in frozen")

    model_filename = os.path.join('.', 'optimized_carc34.pb')
    with gfile.FastGFile(model_filename, 'rb') as f:
      graph_def = tf.GraphDef()
      graph_def.ParseFromString(f.read())
      ##for n in graph_def.node:
      ##  print (n.name)
      output_tensor = tf.import_graph_def(graph_def, name='', return_elements=[MODEL_OUTPUT_TENSOR_NAME])
      print ("Found in optimized")
```

**Comment by arun-kumark** [comment]
Hi Xiusir
I need you help, I am facing the same issue while running the optimize_for_inference.py script.
My arguments are as below:

python optimize_for_inference.py --input=res101_faster_rcnn_iter_5600.pb --output=optimized_graph.pb --input_names="input" --output_names="final_result"

My pb file is based on Resnet-101. I am sharing my res101_faster_rcnn_iter_5600.ckpt.custom file below to see it's contents:

`node {
name: "Placeholder"
op: "Placeholder"
attr {
key: "dtype"
value {
type: DT_FLOAT
}
}
attr {
key: "shape"
value {
shape {
dim {
size: 1
}
dim {
size: -1
}
dim {
size: -1
}
dim {
size: 3
}
}
}
}
}
node {
name: "Placeholder_1"
op: "Placeholder"
attr {
key: "dtype"
value {
type: DT_FLOAT
}
}
attr {
key: "shape"
value {
shape {
dim {
size: 1
}
dim {
size: 3
}
}
}
}
}
node {
name: "Placeholder_2"
op: "Placeholder"
attr {
key: "dtype"
value {
type: DT_FLOAT
}
}
attr {
key: "shape"
value {
shape {
dim {
size: -1
}
dim {
size: 5
}
}
}
}
}
node {
name: "resnet_v1_101/Pad/paddings"
op: "Const"
attr {
key: "dtype"
value {
type: DT_INT32
}
}
attr {
key: "value"
value {
tensor {
dtype: DT_INT32
tensor_shape {
dim {
size: 4
}
dim {
size: 2
}
}
tensor_content: "\000\000\000\000\000\000\000\000\003\000\000\000\003\000\000\000\003\000\000\000\003\000\000\000\000\000\000\000\000\000\000\000"
}
}
}
}
node {
name: "resnet_v1_101/Pad"
op: "Pad"
input: "Placeholder"
input: "resnet_v1_101/Pad/paddings"
attr {
key: "T"
value {
type: DT_FLOAT
}
}
attr {
key: "Tpaddings"
value {
type: DT_INT32
}
}
}
node {
name: "resnet_v1_101/conv1/weights/Initializer/truncated_normal/shape"
op: "Const"
attr {
key: "_class"
value {
list {
s: "loc:@resnet_v1_101/conv1/weights"
}
}
}
attr {
key: "dtype"
value {
type: DT_INT32
}
}
attr {
key: "value"
value {
tensor {
dtype: DT_INT32
tensor_shape {
dim {
size: 4
}
}
tensor_content: "\007\000\000\000\007\000\000\000\003\000\000\000@\000\000\000"
}
}
}
}
node {
name: "resnet_v1_101/conv1/weights/Initializer/truncated_normal/mean"
op: "Const"
attr {
key: "_class"
value {
list {
s: "loc:@resnet_v1_101/conv1/weights"
}
}
}

I am getting tyhe below error :

_result" /home/arun/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6 return f(*args, **kwds) /home/arun/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype fromfloattonp.floatingis deprecated. In future, it will be treated asnp.float64 == np.dtype(float).type. from ._conv import register_converters as _register_converters Traceback (most recent call last): File "optimize_for_inference.py", line 146, in <module> app.run(main=main, argv=[sys.argv[0]] + unparsed) File "/home/arun/anaconda3/lib/python3.6/site-packages/tensorflow/python/platform/app.py", line 48, in run _sys.exit(main(_sys.argv[:1] + flags_passthrough)) File "optimize_for_inference.py", line 90, in main FLAGS.output_names.split(","), FLAGS.placeholder_type_enum) File "/home/arun/anaconda3/lib/python3.6/site-packages/tensorflow/python/tools/optimize_for_inference_lib.py", line 109, in optimize_for_inference placeholder_type_enum) File "/home/arun/anaconda3/lib/python3.6/site-packages/tensorflow/python/tools/strip_unused_lib.py", line 83, in strip_unused raise KeyError("The following input nodes were not found: %s\n" % not_found) KeyError: "The following input nodes were not found: {'input'}\n"
Please help me identifying what arguments you passed to run the script.

python touch.py -m res101_faster_rcnn_iter_5600.ckpt.meta -o res101_faster_rcnn_iter_5600.pb -b

and my touch.py has following contents, which I took from some repository:

#! /usr/bin/python3
#generate a pb file !
from __future__ import print_function
import tensorflow as tf
import argparse
import os

if __name__ == '__main__':


    parser = argparse.ArgumentParser()
    parser.add_argument('-m','--model',required=True, type=str, metavar='' ,help="model meta graph AKA .ckpt.meta file to start with")
    parser.add_argument('-o', '--output_file', type=str, required=True, metavar='',help="Where to save the output file ")
    parser.add_argument('-b','--bin',dest='binary_file',action='store_true', help="saves file as binary instead of default text file")
    parser.add_argument('-v','--verbose',dest='verbosity',action='store_true',help="trigger the verbose output (show tensorflow debug info)")
    args = parser.parse_args()
    #get output dir from output file name
    output_dir = os.path.dirname(args.output_file)
    if output_dir == "":
        output_dir="."
    output_file_name = os.path.basename(args.output_file)
    #setting verbosity    
    os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
    if args.verbosity:
        os.environ['TF_CPP_MIN_LOG_LEVEL'] = '0'
    print ("Welcome to tensorflow graph saver")
    print ("loading ",os.path.basename(args.model))
    #opening a new session
    with tf.Session() as sess:
        new_saver = tf.train.import_meta_graph(args.model)
        new_saver.restore(sess,args.model.replace(".meta","") )
        print ("model",os.path.basename(args.model),"loaded correctly... converting it to graph")
        tf.train.write_graph(sess.graph, output_dir ,output_file_name, as_text=not args.binary_file)
        print("conversion done, output file saved as",args.output_file)
